src/Compilers/Core/Portable/Emit/CommonPEModuleBuilder.cs

```diff
@@ -728,11 +728,41 @@ private void EnsurePrivateImplementationDetailsStaticConstructor(PrivateImplemen
         /// </summary>
         private sealed class SynthesizedDefinitions
         {
-            public ConcurrentQueue<Cci.INestedTypeDefinition> NestedTypes;
+            private ConcurrentQueue<Cci.INestedTypeDefinition> NestedTypes;
             public ConcurrentQueue<Cci.IMethodDefinition> Methods;
             public ConcurrentQueue<Cci.IPropertyDefinition> Properties;
             public ConcurrentQueue<Cci.IFieldDefinition> Fields;
 
+            // Nested types may be queued from concurrent threads, but we need to emit them
+            // in a deterministic order.
+            internal IEnumerable<Cci.INestedTypeDefinition> DeterministicNestedTypes
+            {
+                get
+                {
+#if DEBUG
+                    if (NestedTypes is not null)
+                    {
+                        foreach (var group in NestedTypes.GroupBy(t => t.Name, StringComparer.Ordinal))
+                        {
+                            var arities = group.Select(g => g.GenericParameters.Count());
+                            Debug.Assert(arities.All(a => a == arities.First()));
+                        }
+                    }
+#endif
```

Comment:
Can the assert be simplified (or at least documented)?

Perhaps:
```
Debug.Assert(NestedTypes is null ||
    NestedTypes.GroupBy(t => t.Name, StringComparer.Ordinal).All(g => g.Count() == 1));
```

Reply:
`Distinct` returns an enumerable...
But we can indeed compare counts.